### PR TITLE
[untested]: Make walker resilient to dir access errors

### PR DESCRIPTION
**Pull request by scruffystuffs**
# Overview

When the discovery traverses the file tree, it crashes if there are any errors reading the directory, or executing the directory handler.

This PR only addresses the read errors, since handler errors could be valid.

## Acceptance criteria

Walker is resilient to errors during the `listdir` operation.

## Testing plan

TBD

## Risks

We lose information here, since we can't log the directories that failed (too noisy to do it for every discovery).

## References

Internal slack thread: https://teamfossa.slack.com/archives/C0155DTGWB1/p1655439461261949

## Checklist

- [ ] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [ ] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [ ] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

### CODE REVIEW

File: src/Discovery/Walk.hs
Comment (periodic) [inline src/Discovery/Walk.hs:174]: Any reason not to use `fromMaybe ([], []) <$> recover $ listDir curdir`?  It doesn't seem to be a firm choice in our style guide, but I like that it makes the read order a bit more consistent instead of having to jump way over to the `fmap` to see how that affects the `fromMaybe`.

**Review by periodic — APPROVED** [review]
LGTM.

As for a testing plan, you could probably write a test that revokes list permission from a directory, runs the test, then gives back any permissions required for clean-up, if any.

File: src/Discovery/Walk.hs
Comment (scruffystuffs) [reply src/Discovery/Walk.hs:174]: I think it makes it easier to read since there's no operator confusion this way, but with `a <$> b $ c` it can sometimes not be clear where the functor operation reaches to.  This particular case is a bad example, but this pattern chains well.

Comment (scruffystuffs) [reply src/Discovery/Walk.hs:174]: Also, I think that functions like `fmap` aren't really interesting in terms of reviewing.  If the types are strict, and there's only one way a function could work with the right type, I just generally assume that it's laid out right.  If we had to check that kind of thing anyway, what's the point of having the strict types?

Comment (periodic) [reply src/Discovery/Walk.hs:174]: I don't understand what you mean about "aren't really interesting in terms of reviewing".

I'm just pointing out that as a reader this line is a little awkward because it mixes prefix and infix operators. I think leaning more towards parentheses or more towards infix operators would make it easier to read and understand at a glance.

Comment (scruffystuffs) [reply src/Discovery/Walk.hs:174]: Actually, looking closer at this, there's a good argument for making this pattern a diagnostic helper.

Comment (periodic) [reply src/Discovery/Walk.hs]: Sorry, I mispoke.  When said we were mixing operators I meant mixing infix and prefix functions.  It's not really important for my point.

I'm trying to focus on what is easiest to read and understand at a glance.  I found this line harder to read than most and I think we could do something to make it easier to read.  That could be using conventions like `<$>` (which I may just be personally comfortable with), it may be using multiple lines, it may be refactoring the pattern into a helper (as you suggested).  All I'm suggesting is that you could probably make this easier to read before merging, but I'm not blocking on it.

Though I'm confused on one point. When you say "interesting to a reviewer", do you mean me as the code reviewer, or do you mean a general reader reviewing the code?

Comment (scruffystuffs) [reply src/Discovery/Walk.hs]: reviewer ~ code reviewer

Also, I made it easier by adding helpers to diagnostics.

**Pull request closed without merging**
